Fix differences between release and debug builds

jamii opened the pull request:
## hash_log

From the module docs:

```
A tool for narrowing down the point of divergence between two executions that should be identical.
Sprinkle calls to `emit(some_hash)` throughout the code.
With `-Dhash-log-mode=create`, all emitted hashes are written to ./hash_log.
With `-Dhash-log-mode=check`, all emitted hashes are checked against the hashes in ./hash_log.
Otherwise, calls to `emit` are noops.
```

At the moment the only calls to `emit` are on read/write start/finish in test storage. This is pretty high leverage though, since most decisions end up being committed to disk at some point.

I added a single hash_log run with a low event count to CI to act as a regression test.

## Zero grid blocks

Zero grid blocks on init and before every point of reuse. 

This fixes several buffer bleeds, including https://github.com/tigerbeetledb/tigerbeetle/issues/404. 

We also weren't initializing filter blocks at all so they all converged towards being all set. I don't see any perf improvements from fixing this but our one benchmark atm only stresses compaction, not lookups.

In some places we used to set blocks to `undefined`, but this appears to have different behavior in debug vs release so I changed these all to zero too.

I also noticed https://github.com/tigerbeetledb/tigerbeetle/issues/419 while poking around in the grid.

## Zero filter padding

We explicitly set the padding to zero for index and data blocks, but not for filter blocks. Shouldn't actually be necessary to zero padding for any of them now, but better safe than sorry.

## Zero messages

These were set to undefined in debug builds. Given the issues above, I switched this to zero. I also switched an assert about zeroed padding that only ran in debug builds - it should pass in release builds now too. (If we're worried about perf we could put it behind `config.verify` instead - I mostly just wanted to remove debug vs release differences).

jamii committed:
Add a tool for finding differences between test runs that should be identical.

jamii committed:
hash_log the contents of every read and write on test/storage

jamii committed:
Improve hash_log error messages.

jamii committed:
Zero grid blocks on alloc

jamii committed:
Zero cache blocks before reuse

jamii committed:
Zero padding in filter blocks

jamii committed:
Zero cache blocks before reuse in read_iop

jamii committed:
Always zero message buffers.

jamii committed:
Add hash_log regression test to CI

sentientwaffle commented on .github/workflows/test_and_build.yml:
```diff
@@ -151,6 +151,17 @@ jobs:
 
+  # This both checks that the hash_log builds and acts as a regression test for 
+  # https://github.com/tigerbeetledb/tigerbeetle/issues/404
+  fuzz_hash_log:
```
Can you add this to the `bors_tigerbeetle` job?

jamii replied:
I see, bors doesn't run all jobs?

sentientwaffle replied:
Right, I think `bors_tigerbeetle` determines which jobs must pass for bors to consider the PR to pass.

jamii committed:
Instead of zeroing unused space in table_builder, assert in compaction that blocks are already zeroed.

Also add an extra state to BlockWrite so that we can assert that writes have finished before writing new data to the block.

sentientwaffle commented:
bors r+
